cdap-security/src/main/java/co/cask/cdap/security/guice/SecureStoreModules.java

```diff
@@ -66,43 +86,103 @@ public final Module getDistributedModules() {
-        bind(SecureStoreManager.class).toProvider(new TypeLiteral<StoreProvider<SecureStoreManager>>() { });
+        bind(SecureStore.class).toProvider(new TypeLiteral<DistributedStoreProvider<SecureStore>>() { });
+        bind(SecureStoreManager.class).toProvider(new TypeLiteral<DistributedStoreProvider<SecureStoreManager>>() { });
       }
     };
   }
 
   @Singleton
-  private static final class StoreProvider<T> implements Provider<T> {
+  private static final class DistributedStoreProvider<T> implements Provider<T> {
     private final CConfiguration cConf;
     private final Injector injector;
 
     @Inject
-    private StoreProvider(final CConfiguration cConf, Injector injector) {
+    private DistributedStoreProvider(final CConfiguration cConf, Injector injector) {
       this.cConf = cConf;
       this.injector = injector;
     }
 
     @Override
     @SuppressWarnings("unchecked")
     public T get() {
-      if (!SecureStoreUtils.isKMSBacked(cConf)) {
-        throw new IllegalArgumentException("File based secure store is not supported in distributed mode. " +
+      boolean kmsBacked = KMS_BACKED.equalsIgnoreCase(cConf.get(Constants.Security.Store.PROVIDER));
+      if (kmsBacked && isKMSCapable()) {
+        try {
+          return (T) injector.getInstance(Class.forName("co.cask.cdap.security.store.KMSSecureStore"));
+        } catch (ClassNotFoundException e) {
+          // KMSSecureStore could not be loaded
+          throw new RuntimeException("CDAP KMS classes could not be loaded. " +
+                                       "Please verify that CDAP is correctly installed");
+        }
+      }
+      if (kmsBacked) {
+        throw new IllegalArgumentException("Could not find classes required for supporting KMS based secure store. " +
+                                             "KMS backed secure store depends on " +
+                                             "org.apache.hadoop.crypto.key.kms.KMSClientProvider being available. " +
+                                             "This is supported in Apache Hadoop 2.6.0 and up and on " +
+                                             "distribution versions that are based on Apache Hadoop 2.6.0 and up.");
+      }
+
+      if (FILE_BACKED.equalsIgnoreCase(cConf.get(Constants.Security.Store.PROVIDER))) {
+        throw new IllegalArgumentException("Only KMS based provider is supported in distributed mode. " +
                                              "To be able to use secure store in a distributed environment you" +
-                                             "will need to use the Hadoop KMS based provider.");
+                                             "will need to use the Hadoop KMS based provider." +
+                                             "For more details please refer to " +
+                                             "https://wiki.cask.co/display/CE/Using+Secure+Store.");
```

Comment:
Nope.

> On Aug 2, 2016, at 3:34 PM, Bhooshan Mogal <notifications@github.com> wrote:
> 
> In cdap-security/src/main/java/co/cask/cdap/security/guice/SecureStoreModules.java <https://github.com/caskdata/cdap/pull/6293#discussion_r73250855>:
> 
> >                                               "To be able to use secure store in a distributed environment you" +
> > -                                             "will need to use the Hadoop KMS based provider.");
> > +                                             "will need to use the Hadoop KMS based provider." +
> > +                                             "For more details please refer to " +
> > +                                             "https://wiki.cask.co/display/CE/Using+Secure+Store.");
> there isn't any internal information in this wiki, right?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/caskdata/cdap/pull/6293/files/7214cc401842ba8ee0980ad47dea81a2c146ba6e#r73250855>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAFGk6wuLxMOfBqNzixDDE2Gdmn9RRluks5qb8XYgaJpZM4Ja37Y>.
>